### PR TITLE
Allow exposing caffe2 operators with variable number of input tensors to c10

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#17491 Allow exposing caffe2 operators with variable number of input tensors to c10**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14220705/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17522 Allow dispatch based on tensor list args&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14235840/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17579 Fix InputSize/OutputSize for ivalue based operators&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14253094/)

Before, there was no way to expose a caffe2 operator that had a variable number of inputs.
Now, this is allowed by giving the operator one tensor list input.
Note that the tensor list must be the first input, and that any other tensor inputs will be ignored and inaccessible in this case.

Differential Revision: [D14220705](https://our.internmc.facebook.com/intern/diff/D14220705/)

### CODE REVIEW

File: torch/csrc/jit/register_c10_ops.cpp [hunk 1]
```diff
@@ -16,11 +23,13 @@ Operator createOperatorFromC10(const c10::OperatorHandle& op) {
       for (auto iter = stack.end() - input_size; iter != stack.end(); ++iter) {
         // TODO Remove the .defined() check once we don't have undefined tensors on the stack anymore (@wanchaol is working on this)
         if (iter->isTensor() && iter->toTensor().defined()) {
-          at::Tensor tensor = std::move(*iter).toTensor();
-          if (tensor.requires_grad()) {
-            throw std::runtime_error("Autograd not yet supported for c10 ops.");
+          *iter = unwrap(std::move(*iter).toTensor());
+        }
+
+        if (iter->isTensorList()) {
```
Comment (ezyang) [inline torch/csrc/jit/register_c10_ops.cpp]: else if?

File: torch/csrc/jit/register_c10_ops.cpp [hunk 2]
```diff
@@ -16,11 +23,13 @@ Operator createOperatorFromC10(const c10::OperatorHandle& op) {
       for (auto iter = stack.end() - input_size; iter != stack.end(); ++iter) {
         // TODO Remove the .defined() check once we don't have undefined tensors on the stack anymore (@wanchaol is working on this)
         if (iter->isTensor() && iter->toTensor().defined()) {
```
Comment (ezyang) [inline torch/csrc/jit/register_c10_ops.cpp]: If your invariant is that Tensors must all come first, shouldn't you break as soon as you reach the first non-tensor argument?

**Review by ezyang — APPROVED** [review]
approved with comments

File: torch/csrc/jit/register_c10_ops.cpp (hunk 2 above)
Comment (smessmer) [reply torch/csrc/jit/register_c10_ops.cpp]: That's the invariant for wrapped caffe2 operators. Native c10 operators don't have this restriction.

**Pull request closed without merging**